### PR TITLE
Standalone content-tag implemented via conditional exports in package.json

### PR DESCRIPTION
**Pull request by NullVoxPopuli**
Alternative to: https://github.com/embroider-build/content-tag/pull/41

Changes:
- there are now two wasm-pack builds, one targeting noed, and one "browser" or "standalone" (as browser targets tend to be "standalone" / runnable anywhere) -- if we didn't need to support CJS, we could get away with _only_ the standalone build
- the root package.json is now _the_ package.json that is published
  - now has name, description, etc
  - `files` entry includes the `pkg` directory
  - has `exports` with conditional/subpath entries to target node/browser separately since wasm-pack generates _very_ different output for each target (there isn't a need for this if they went for full async-only API, but :shrug: )
- To test that the browser changes actually work, vite has been added to devDependencies as well as an index.html
  - I tried to automate browser testing with vitest's browser mode: https://vitest.dev/guide/browser#browser-option-types -- but the browser mode does not support the same work around that `vite` has for https://github.com/vitejs/vite/issues/9731 -- which is setting `resolve-alias` for `"package-name": "."` (current directory) so that vite can resolve the "self's" exports.  So we currently don't have automated browser testing atm.  We could set up qunit instead of vitest, but we currently need to handroll a reporter to the CLI, because vite/vitest haven't given us those APIs to build custom reporters. GlimmerVM is already doing this, so it "just" needs to be extracted.
  This is the error:
    ```
    [plugin:vite:import-analysis] Failed to resolve import "content-tag" from "test-browser/process.test.js". Does the file exist?
    /home/nvp/Development/OpenSource/emberjs/content-tag/test-browser/process.test.js:2:48
    1  |  import { __vi_inject__ as __vi_esm_0__ } from 'vitest'
    2  |  import { __vi_inject__ as __vi_esm_1__ } from 'content-tag'
       |                                                 ^
    3  |  
    4  |
    ```
  - Due to not being able to correctly test with vitest, I added `publint` and `@arethetypeswrong/cli` to help me be more confident that the package.json was correctly configured so we don't have to rapid publishes to fix bugs with package.json config.
    `@arethetypeswrong/cli` reported some issues which are resolved now.
    One being [this issue](https://github.com/arethetypeswrong/arethetypeswrong.github.io/blob/main/docs/problems/FalseESM.md) which is resolved by using node newer than 16 -- which... we don't declare node support at all, and node 16 is EOL'd so it seems fine to bump the CI's node.

--------------------

### In a real app

because my confidence in non-monorepo layouts is _way_ lower (can't test with real projects with real module resolution),

here is a real usage:

- branch with dist assets committed: https://github.com/NullVoxPopuli/content-tag/tree/non-monorepo-dist
- PR to my repl here: https://github.com/NullVoxPopuli/limber/pull/1546
  Results:
  - Works:
    ```ts
    let processor: any;
    let fetchingPromise: Promise<any>;
    
    export async function preprocess(input: string, name: string): Promise<string> {
      if (!fetchingPromise) {
        fetchingPromise = import('content-tag');
      }
    
      if (!processor) {
        let { Preprocessor } = await fetchingPromise;
    
        processor = new Preprocessor();
      }
    
      return processor.process(input, `${name}.js`);
    }
    ```
  - Does not work: 
    ```ts
    import { Preprocessor } from 'content-tag';
    
    const processor = new Preprocessor();
    
    export async function preprocess(input: string, name: string): Promise<string> {
      return processor.process(input, `${name}.js`);
    }
    ```
    but this is due to: 
    https://github.com/embroider-build/embroider/issues/1678
    and this code likely _would_ work once that 1678 issue is resolved.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -1,20 +1,45 @@
...
     "type": "git",
     "url": "git@github.com:embroider-build/content-tag.git"
   },
+  "files": [
+    "pkg"
+  ],
+  "exports": {
+    ".": {
+      "types": "index.d.ts",
+      "browser": {
```
Comment (NullVoxPopuli) [inline package.json:17]: this conditional export is what enables browser mode

Comment (NullVoxPopuli) [reply package.json:17]: (assuming vite isn't falling back to the "default" exports (pkg/node/content_tag.cjs) and polyfilling all the node stuff for the browser)

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -34,10 +34,13 @@ jobs:
...
-    - run: wasm-pack build --target nodejs
+    - name: Install wasm-opt
+      run: cargo install wasm-opt
     - uses: actions/setup-node@v3
       with:
-        node-version: 16
+        node-version: 18
         cache: 'npm'
-    - run: npm ci
+    - run: ./build.sh
```
Comment (void-mAlex) [inline .github/workflows/ci.yml:43]: would be nice to have this consistent with release workflow

Comment (NullVoxPopuli) [reply .github/workflows/ci.yml:43]: the release workflow already run `npm run build`, so this is covered

**Comment by void-mAlex** [comment]
yes, if this did the same thing meaning npm run build the two would be in
sync

**Review by ef4 — APPROVED** [review]

**Pull request merged**